### DISCUSSION TITLE
Error running distributed mesh

### DISCUSSION THREAD

**Discussion by bhollra** [idaholab/moose#18583] [Q&A General] [answered]
Hi and thanks for reading,

I am attempting to run a simulation with a distributed mesh by following the information [here](https://mooseframework.inl.gov/syntax/Mesh/splitting.html). I am able split the mesh successfully with this command: `sam-opt -i input.i --split-mesh 6` creating a folder `./mesh_file.cpr/6/` as I expected.

However, when I attempt to run the simulation on the distributed mesh with `mpiexec -n 6 ~/projects/SAM/sam-opt -i input.i --use-split` I get the following error:

`ERROR: Neither one of the following files can be located:`
`        '/path/./mesh_file.cpr/1/header.cpr' nor`
`        '/path/./mesh_file.cpr'`
`If you are running a parallel job, double check that you've created a split for 1 ranks.`
`Note: One of paths above may refer to a valid directory on your system, however we are attempting to read a valid header file.`

In the documentation it sounds like the code should automatically search for a mesh split by the number of cores specified in the mpiexec command, in this case 6. But this error message seems to indicate the code is looking for a mesh with 1 split. 

I am curious if anyone has any insight into the cause of this error? Is it possible this is a SAM specific issue? 

P.S. My framework information is:
`MOOSE Version:           git commit 48fe948cda on 2020-02-17`
`LibMesh Version:         3b8296b3b7c2b526d0be242fa6208750c0ede69b`
`PETSc Version:           3.11.4`
`SLEPc Version:           3.11.0`

**Comment by GiudGiud** [comment]
Hello

The fact that it's looking for a 1-split makes me think something went terribly wrong with MPI. Like the code was compiled with a different MPI implementation than the mpirun you are using, and each process is identifying the number of ranks incorrectly. 

I'm not familiar with SAM. I dont think they would have modified this, even though their typical use case should not require split meshes. 
@travismui is there anything we should be aware of with regards to SAM and distributed runs?

Do you really need distributed meshes with a pre-split? The impact on performance is usually negligible. It's only done to save some memory (and save some time on IO with slow file systems for the split). What is your application? SAM is more than reasonable with memory usage

Guillaume

**Reply by bhollra** [reply]
Hi Guillaume,

Thanks for your response. In this case, I am using SAM to model a large 3D solid heat conduction domain, so the memory demands are quite large. Without a distributed mesh I am currently limited to running on only one processor. 

My SAM/MOOSE version was about a year and a half old. After updating to the newest version last night, the simulation is now able to load the `'/path/./mesh_file.cpr/6/header.cpr'` folder correctly. So it seems you were right and there likely was some incompatibility issue going on with MPI. 

However, I am now seeing a new error:
`*** ERROR ***`
`The following error occurred in the object "MOOSE Problem", of type "FEProblem".`
`Need Exodus reader to restart variables but the reader is not available`
`Use either FileMesh with an Exodus mesh file or FileMeshGenerator with an Exodus mesh' 'file and with use_for_exodus_restart equal to true`

Included bedlow are my mesh blocks and variable blocks:
`[Mesh]`
`	file = ${meshFileName}`
`[]`

`[Variables]`
`	[./T_solid]`
`		order = FIRST`
`		family = LAGRANGE`
`		block = 'all blocks'`
`		scaling = 1.0E-2`
`		initial_from_file_var = T_solid`
`		initial_from_file_timestep = 2`
`	[../]`
`[]`

It appears to me that the error is cause by the `initial_from_file_var` parameter in the variable sub-block. When `initial_from_file_var` and `initial_from_file_timestep` are removed and replaced with test constant IC:
`[./InitialCondition]`
`type = ConstantIC`
`value = 1000`
`[../]`
The simulation is able to run successfully. 
It seems that variable data are not stored in the split mesh files. Is it possible to initialize a distributed simulation with a variable from an Exodus file that is not distributed via another method? 

Some background on my simulations: The simulation that is currently giving me an error is a transient simulation that needs to be initialized with a temperature distribution calculated by a steady state simulation. The steady state simulation uses multiapps to couple the 3D solid domain with several 1D fluid components. It is my understanding that is is not possible to use multiapps and a distributed mesh together, so I was forced to use one processor for the steady state simulation. Now, attempting the transient simulation, the model no longer includes the 1D fluid multiapps (it is a pure 3D solid conduction problem), so I am hoping to use to distributed mesh to decrease the simulation time. 

Thank you,
Brent

P.S. I have created a very simplified version of my input and reproduced the same errors using the combined-opt moose module, so I feel confident this is not a SAM specific error.
P.P.S. Updated framework information:
`MOOSE Version:           git commit 71aebe45e9 on 2021-08-11`
`LibMesh Version:         `
`PETSc Version:           3.15.1`
`SLEPc Version:           3.15.1`

**Comment by GiudGiud** [comment]
Hello

The solution for your issue is to use a file mesh generator (in one word) and set the parameter ‘use_for_exodus_restart = true’. 

Guillaume

**Reply by bhollra** [reply]
Thanks Guillaume, but when I try this I get the same error message. 
To test this I took the following steps:
1. Update mesh block to the following:
`[Mesh]`
`#       file = ${meshFileName}`
`[./fmg]`
` type = FileMeshGenerator`
` file = mesh.e`
` use_for_exodus_restart = true`
`[../]`
`[]`

2. Create split mesh:
`mpiexec -n 6 ~/MOOSE_08_2021/moose/modules/combined/combined-opt -i simplified.i --split-mesh 6 --split-file mesh_fmg.cpr`

3. Run simulation:
`mpiexec -n 6 ~/MOOSE_08_2021/moose/modules/combined/combined-opt -i simplified.i --use-split --split-file mesh_fmg.cpr`

If it is any help, you can find the full input file here: 
[simplified.txt](https://github.com/idaholab/moose/files/6985289/simplified.txt)

**Comment by GiudGiud** [comment, marked as answer]
Hello

I seem to have missed a few things in my previous reply, I ll try to be more exhaustive here. 

- For your use case, you do need to split the mesh. You can simply use the `--distributed-mesh` option, as it will distribute the mesh among each process and reduce the memory cost of the simulation accordingly. Splitting meshes is used on multi-node jobs to reduce I/O, so each node does not have to load the entire mesh but just the split it is in charge of. 

Splitting meshes is also useful if the mesh does not fit on the memory of a single node. This is not your case it seems. Could you please share with us the size of the problem, eg the number of elements and the number of dofs/element?

- Splitting the mesh, step 2, only creates the mesh split. It does not load the variables and initialize them. So essentially, it prevents you from using the exodus restart capability. The exodus file containing the initial variable values is never loaded, because the mesh split is used instead. I'll let others pitch in on this, but I think we will add a log to document this issue. 

You could work with Checkpoint restart to create the initialization if you absolutely want to use split meshes. This will require using the same amount of nodes, and may fail if the Steady simulation system is not exactly the same as the Transient one. I would just recommend you avoid using split meshes and use the distributed option instead, which can work in conjunction with Exodus restart.

Best
Guillaume

**Reply by bhollra** [reply]
Hi Guillaume,

I really appreciate your continued help on this issue. I had somehow missed the availability of the "--distributed-mesh" command line option. It appears I was making this much more complicated than it needed to be. I am now able to run my simulation distributed across 5 cores. 

If I attempt to use more cores the simulation is either unable to converge or terminates after computing the initial residual of time step one (potentially due to a temporary spike in RAM usage). I have not spent much time troubleshooting this, but running on 5 processors should be good enough for my simulation. 

P.S. 
For some information on the size of the simulation I am attempting please see the following output. This simulation uses between 60-70% of the RAM on my 24GB desktop:
>  `Mesh: `
    `Parallel Type:           distributed`
    `Mesh Dimension:          3`
    `Spatial Dimension:       3`
    `Nodes:                   `
      `Total:                 2805425`
      `Local:                 466047`
      `Min/Max/Avg:           422088/496784/467570`
    `Elems:                   `
      `Total:                 4329440`
      `Local:                 696246`
      `Min/Max/Avg:           663922/757444/721573`
    `Num Subdomains:          727`
    `Num Partitions:          6`
    `Partitioner:             parmetis`

>  `Nonlinear System:`
    `Num DOFs:                2805425`
    `Num Local DOFs:          466047`
    `Variables:               "T_solid" `
    `Finite Element Types:    "LAGRANGE" `
    `Approximation Orders:    "FIRST" `

 > `Auxiliary System:`
    `Num DOFs:                886420`
    `Num Local DOFs:          152121`
    `Variables:               "power_density" `
    `Finite Element Types:    "LAGRANGE" `
    `Approximation Orders:    "FIRST" `

  >`Execution Information:`
    `Executioner:             Transient`
    `TimeStepper:             FunctionDT`
    `Solver Mode:             NEWTON`
    `PETSc Preconditioner:    gamg `
    `MOOSE Preconditioner:    SMP`

**Reply by GiudGiud** [reply]
The inability to converge with more cores is usually just a matter of tuning the tolerances and the pre-conditioner settings. For example, the asm preconditioner will need to have an overlap parameter specified to perform better with more core. Less efficient pre-conditioning can mean that allowing for more linear iterations helps the solve.

With the Newton solve method, it's likely your memory issue is due to the need to store the system matrix. You could switch to the PJFNK solve method. This will not converge as well, but scales better in terms of memory. 

We are 1-2 weeks away from merging new memory profiling utilities in MOOSE (see #16298) if you ever want to determine the cause of the memory spike for sure.

**Reply by bhollra** [reply]
Hi Guillaume,

Another question just came up, this time regarding the compatibility of using `--recover` and `--use-distributed` together.

My desktop had an issue that cause the simulation to terminate early. When I attempted running `mpiexec -n 5 ~/SAM_08_2021/SAM/sam-opt -i  square_full_core_SBO.i --recover --distributed-mesh` I got the following error:

`*** ERROR ***`
`ERROR: Neither one of the following files can be located:`
`        'square_full_core_SBO_out_cp/0270_mesh.cpr/1/header.cpr' nor`
`        'square_full_core_SBO_out_cp/0270_mesh.cpr'`
`If you are running a parallel job, double check that you've created a split for 1 ranks.`
`Note: One of paths above may refer to a valid directory on your system, however we are attempting to read a valid header file.`

I found this interesting because when I attempted to run the input with a restart block added:
`[Problem]`
`restart_file_base = square_full_core_SBO_out_cp/LATEST`
`[]`
The simulation has no issues initializing using the checkpoint. To confirm:
`*** Info ***                                                                                                        `
`Using /home/bhollra/spring_2021/summer_2021/square_full_core/SBO/square_full_core_SBO_out_cp/0270 for restart.      `
is output in the terminal.

There are workarounds I can use to reinitialize my simulation using the results from the last output time step of the terminated simulation, but I did find it interesting that using `--recover` did not work while using a restart from the same checkpoint did.

**Reply by GiudGiud** [reply]
Interesting. This might be a bug with the --recover option and splits. I ll try to reproduce that later and I ll raise an issue if needed. 
Thanks for letting us know. 

I'm glad it works with the restart option though.

**Reply by GiudGiud** [reply]
Ok I confirmed recover+split worked on the simple diffusion case. 
I simply made that case a transient and added
`  checkpoint = true
`

Then I ran:
```
../../../moose_test-opt -i simple_diffusion.i --split-mesh 4 --split-file mesh_split
mpirun -n 4 ../../../moose_test-opt -i simple_diffusion.i --use-split 4 --split-file mesh_split --half-transient
mpirun -n 4 ../../../moose_test-opt -i simple_diffusion.i --use-split 4 --split-file mesh_split --recover
```

I would recommend you make it work on a small example like this before trying to see what is different with your case.

**Reply by GiudGiud** [reply]
You could also have a look at the recover/checkpoint tests in test/tests/mesh/checkpoint btw
